### ISSUE TITLE
KMS ENV Vars

### ISSUE THREAD

**Issue by DaveBlooman** [apex/apex#651]
The KMS ARN that is used by Lambda only encrypts secrets at REST, this means that while they are viewable in the AWS console.  The better approach would be to have a way to encrypt the secrets before hand, then upload them in the encrypted state.

This would mean that the lambda function would have to know which env vars are encrypted and which aren't, but I think this could be handled either by the shim, or auto generating the code needed for each language by reading the env vars from the function.json file.  We could also have a flag on each ENV var that is a boolean for encrypted or not.

```json
{
  "name": "fop",
  "description": "bar",
  "memory": 128,
  "timeout": 1,
  "environment": [
    {
      "someAPIkey": "823f98wejf9asj0ecjes09fjs0ef",
      "encrytpted": true
    },
    {
      "someotherAPIkey": "awdadwdawdawdaw",
      "encrytpted": true
    }
  ],
  "role": "arn:aws:iam::00000000000:role/lambda_function",
  "kms_arn": "arn:aws:kms:eu-west-1:000000000000:key/d23234233-9999-9999-9990-9349349343943"
}

```

This would change is breaking though, so as an initial step, would could just have people encrypt everything.  

```json
{
  "name": "fop",
  "description": "bar",
  "memory": 128,
  "timeout": 1,
  "environment": {
    "someAPIkey": "823f98wejf9asj0ecjes09fjs0ef",
    "someotherAPIkey": "awdadwdawdawdaw"
  },
  "environment_encrytpted": true,
  "role": "arn:aws:iam::00000000000:role/lambda_function",
  "kms_arn": "arn:aws:kms:eu-west-1:000000000000:key/d23234233-9999-9999-9990-9349349343943"
}
```

This approach assumes that you are using plain text values, not values already encrypted.  I think this is the more sensible approach as if you have already encrypted the values, then you probably don't have a problem committing them to VCS.

**Comment by tj** [comment]
Related: #613, and #532 would provide an alternative allowing `env(NAME)`, so you could put the vars in .envrc or similar.

**Comment by DaveBlooman** [comment]
@tj Those solutions wouldn't solve the underlying issue of encrypting on AWS.  When env vars were introduced, the idea was that everything was encrypted at REST using the standard KMS token.  

The feature I'm looking for is to actually encrypt before leaving the host machine.  The benefit there would be that Apex could handle the encrypt via KMS and then the function would be given the right IAM to decrypt.  I'm working on a PR now which will just extend the variable mapping and if the flag is set, will encrypt all variables.

My setup is to pull KV out of a Consul sever and then with a script, use the command line flags to add ENVs, but they are not encrypted once they leave the Consul server.

**Comment by tj** [comment]
Yep just wanted to mention the other features for people browsing. I haven't looked into it much, it would be great if we can encrypt client-side with the KMS defaults, adding the ARN and IAM perms for everyone would be a little annoying

**Comment by DaveBlooman** [comment]
You will need a KMS ID at the very least, which will be account specific. There is no default key as such, so as long as people provide that, we can encrypt.  

I have a PR that is nearly ready for review that will show one approach for this.

**Comment by tj** [comment]
Lambda provides a default, I'm not sure we can grab the info for it, sounds like you maybe have to supply your own in those scenarios

**Comment by DaveBlooman** [comment]
Made #652 if we go with the idea that all variables are encrypted.  

Been trying to think about the scoping for this with project settings.  I doubt you would want all lambda vars encrypted on every function, so function only makes sense.  

We could also exclude default lambda variables from encryption, which shouldn't be that difficult, but would mean that any user defined var in the function would be encrypted. 

Any initial feedback?

**Comment by kadrach** [comment]
I don't fully understand this advertised feature. From the documentation:

> Apex now supports the new native AWS Lambda environment variables, which are encrypted via KMS. There are several ways to set these variables, let’s take a look!

I can see in the AWS console that neither are my apex-deployed variables encrypted, nor is the encryption helper flag activated. This causes the AWS supplied code to error on module initialization:

```
import boto3
import os
from base64 import b64decode

ENCRYPTED = os.environ['CLIENT_SECRET']
DECRYPTED = boto3.client('kms').decrypt(CiphertextBlob=b64decode(ENCRYPTED))['Plaintext']

...
```
```
module initialization error: Incorrect padding
```
Is there a second KMS encryption setting/method for AWS Lambda that I am not aware of?

If I manually enable the encryption helper in the console and encrypt the variables in the console, the module initializes correctly (via console > test), but not via `apex invoke`.

**Comment by mda590** [comment]
@kadrach what did you end up doing here? I am running into the exact same problem trying to figure out a way to mark certain variables for encryption within a function.json file. I don't think the feature advertised "Apex now supports the new native AWS Lambda environment variables, which are encrypted via KMS" has been implemented with KMS. Looking through the code for functions, I don't see anywhere that actually uses the "kms_arn" we are providing.

It looks like @DaveBlooman has part of this working here: https://github.com/DaveBlooman/apex/blob/kms_envvars/function/function.go but I am not sure.
